modules/core/src/main/java/org/apache/ignite/internal/processors/service/ServiceInfo.java

```diff
@@ -37,6 +39,9 @@
     /** */
     private static final long serialVersionUID = 0L;
 
+    /** Context. */
+    private final GridKernalContext ctx;
```

Comment:
`ServiceInfo` is a serializable unit.
I'm not sure that there is a sense to transfer `GridKernalContext` instance.

Maybe it's better to make the field transient and initialize it after deserialization.
What do you sink?

Reply:
Ok, I will fix it.

Reply:
We can't get context after deserialization, because we don't know Ignite instance at this stage. I've marked the field as transient and added a check for null to avoid NPE. I think this is the best we can do. The class resolving logic will not work for services with lazy configuration and deployed via deployment SPI, after serialization/deserialization of `ServiceInfo`. But I think it's not such a common case - `ServiceInfo` not serialized by Ignite and serialization of `ServiceInfo` by the user looks strange. I think this class should not be `Serializable` at all.

Reply:
We can handle it in `IgniteServiceProcessor` in places where `registeredServices#put`is called.
Adding to `registeredServices` is the first place when Ignite's node receives `ServiceInfo`

We can introduce a new method:
```
private void registerService(ServiceInfo desc) {
    desc.setContext(ctx);
    registeredServices.put(desc.serviceId(), desc);
}
```

What do you think?

Reply:
I have missed the point. `ctx` already passed to `ServiceInfo` constructor right before `registeredServices#put`. And this works fine, the only problem will show up if the user will serialize and deserialize `ServiceInfo` items in some circumstances. But this problem is not solved by your proposal either.

Reply:
I've tried to cover the case with statically configured services, which will be received during node joining.

In this case server (not user) will deserialize `ServiceInfo` without calling the new constructor.

Look at:
https://github.com/apache/ignite/blob/8f824ca5822d4d7d0b9eb12aa726c32a56776bbc/modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java#L350

I think we need a test for this case too.

Reply:
Thanks, fixed it, and added a check to the test.

I found one more problem with services using classes deployed by deployment SPI: such services can't be deployed to the joined node. It can be reproduced if we change in `testServiceDeploymentViaDeploymentSpi` from `deployClusterSingleton` to `deployNodeSingleton`. We can't register class in deployment SPI before node starts, so, after new node joins it receives services information from the cluster and tries to start service locally, but can't find deployment SPI classes, so a registration of such services fails. This issue is not related to this thicket. It looks conceptual and I don't have any ideas on how to solve it right now. 

Also, looks like deployment SPI is not supported by the old service processor implementation, so I've reverted my changes related to GridServiceProcessor. I don't know why TC bot visa was green, I've double-checked it manually, and test `testServiceDeploymentViaDeploymentSpi` failed. This issue can be easily fixed, if we copy-paste `copyAndInject` method from the new implementation, but since the old implementation is deprecated I think we can live with these known issues.

Reply:
Looks like `desc.context(ctx)` should be called here as well:
1. https://github.com/apache/ignite/blob/1ce95c603e15aa81d665c0be5b2b89466982d0fb/modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java#L389

2. https://github.com/apache/ignite/blob/1ce95c603e15aa81d665c0be5b2b89466982d0fb/modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java#L1519

I'd suggest to introduce a new method:
```
    private void registerService(ServiceInfo desc) {
        desc.context(ctx);
        
        registeredServices.put(desc.serviceId(), desc);
    }
```
It'll allow us to not miss such places in future.

What do you think?

Reply:
Ok, fixed